### PR TITLE
Categories crud - change trait to hierarchy builder

### PR DESCRIPTION
**Pull request by abror9109**

### CODE REVIEW

File: app/Builders/Hierarchy/Hierarchy.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace App\Builders\Hierarchy;
+
+class Hierarchy
```
Comment (AnvarovD) [inline app/Builders/Hierarchy/Hierarchy.php]:
```suggestion
class HierarchyBuilder
```

cs-fixer i zapusk qilin

File: app/Builders/Hierarchy/Hierarchy.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        $this->model = new $model();
+    }
+
+    public function SetParentColumn($parentColumnName)
+    {
+        $this->parentColumnName = $parentColumnName;
+        return $this;
+    }
+
+    public function SetIdNamePair(array $idNameArray)
```
Comment (AnvarovD) [inline app/Builders/Hierarchy/Hierarchy.php]:
```suggestion
    public function SetIdNamePair(int $id, string $name)
```
agar bir nechta bo'sa locale ga qarab oladigan qilamz

File: app/Builders/Hierarchy/Hierarchy.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        $this->id = $idNameArray[0];
+        $this->name = $idNameArray[1];
+        return $this;
+    }
+
+    protected function MakeHierarchicalOptions(&$result = [], $parent_id = null, $level = 0)
```
Comment (AnvarovD) [inline app/Builders/Hierarchy/Hierarchy.php]:
```suggestion
    protected function MakeHierarchicalOptions(array &$result = [], int $parent_id = null,int $level = 0): void
```
Type hinting ga etibor berin

File: app/Builders/Hierarchy/Hierarchy.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        $items = $items->pluck($this->name, $this->id)->toArray();
+
+        foreach ($items as $id => $value) {
+            $result[$id] = " " . str_repeat("-", $level) . " " . $value;
+            $this->MakeHierarchicalOptions($result, $id, $level+1);
+        }
+    }
+
+    public function Make()
```
Comment (AnvarovD) [inline app/Builders/Hierarchy/Hierarchy.php]:
```suggestion
    public function Make(): array
```

File: app/Builders/Hierarchy/Hierarchy.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+namespace App\Builders\Hierarchy;
+
+class Hierarchy
+{
+    private $id;
+    private $name;
+    private $model;
+    private $parentColumnName = "parent_id";
+
+    public function __construct(String $model)
```
Comment (AnvarovD) [inline app/Builders/Hierarchy/Hierarchy.php]:
```suggestion
    public function __construct(string $model)
```

File: app/Models/Category.php [hunk 6]
```diff
@@ -2,19 +2,13 @@
...
 {
-    use HasFactory, HasHierarchyTrait;
-
-    protected static function GetIdValuePair()
-    {
-        return ['title', 'id'];
-    }
+    use HasFactory;
 
     public function children(): HasMany
```
Comment (AnvarovD) [inline app/Models/Category.php]: cs-fixer i zapusk qilin

File: app/Models/Skill.php [hunk 7]
```diff
@@ -2,19 +2,13 @@
...
 {
-    use HasFactory, HasHierarchyTrait;
-
-    protected static function GetParentColumn()
-    {
-        return 'parent_skill';
-    }
+    use HasFactory;
 
     public function children(): HasMany
```
Comment (AnvarovD) [inline app/Models/Skill.php]: cs-fixer i zapusk qilin

File: app/MoonShine/Resources/CategoryResource.php [hunk 8]
```diff
@@ -20,11 +20,15 @@ class CategoryResource extends Resource
 
     public function fields(): array
```
Comment (AnvarovD) [inline app/MoonShine/Resources/CategoryResource.php]: cs-fixer i zapusk qilin

File: app/MoonShine/Resources/SkillResource.php [hunk 9]
```diff
@@ -8,6 +8,7 @@
 use MoonShine\Fields\Select;
 use MoonShine\Resources\Resource;
 use MoonShine\Actions\FiltersAction;
+use App\Builders\Hierarchy\Hierarchy;
 use Illuminate\Database\Eloquent\Model;
 
 class SkillResource extends Resource
```
Comment (AnvarovD) [inline app/MoonShine/Resources/SkillResource.php]: cs-fixer i zapusk qilin

File: app/Builders/Hierarchy/Hierarchy.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        $this->parentColumnName = $parentColumnName;
+        return $this;
+    }
+
+    public function SetIdNamePair(array $idNameArray)
+    {
+        $this->id = $idNameArray[0];
+        $this->name = $idNameArray[1];
+        return $this;
```
Comment (AnvarovD) [inline app/Builders/Hierarchy/Hierarchy.php]:
```suggestion
        $this->id = $id;
        $this->name = $name;
        return $this;
```

File: app/Builders/Hierarchy/Hierarchy.php (hunk 1 above)
Comment (abror9109) [reply app/Builders/Hierarchy/Hierarchy.php]: +

File: app/Builders/Hierarchy/Hierarchy.php (hunk 2 above)
Comment (abror9109) [reply app/Builders/Hierarchy/Hierarchy.php]: ozgartirildi

File: app/Builders/Hierarchy/Hierarchy.php (hunk 3 above)
Comment (abror9109) [reply app/Builders/Hierarchy/Hierarchy.php]: Togirlandi

**Pull request merged**